services/gundeck/test/unit/MockGundeck.hs

```diff
@@ -0,0 +1,712 @@
+
+shrinkNotifs :: [(Notification, [Presence])] -> [[(Notification, [Presence])]]
+shrinkNotifs = shrinkList (\(notif, prcs) -> (notif,) <$> shrinkList (const []) prcs)
+
+
+----------------------------------------------------------------------
+-- monad type and instances
+
+newtype MockGundeck a = MockGundeck
+  { fromMockGundeck :: ReaderT MockEnv (StateT MockState (RandT StdGen Identity)) a }
+  deriving (Functor, Applicative, Monad, MonadReader MockEnv, MonadState MockState, MonadRandom)
+
+runMockGundeck :: MockEnv -> MockGundeck a -> (a, MockState)
+runMockGundeck env (MockGundeck m) =
+  runIdentity . (`evalRandT` mkStdGen 0) $ runStateT (runReaderT m env) mempty
+
+instance MonadThrow MockGundeck where
+  throwM = error . show  -- (we are not expecting any interesting errors in these tests, so we might
+                         -- as well crash badly here, as long as it doesn't go unnoticed...)
+
+instance MonadPushAll MockGundeck where
+  mpaNotificationTTL = pure $ NotificationTTL 300  -- (longer than we want any test to take.)
+  mpaMkNotificationId = mockMkNotificationId
+  mpaListAllPresences = mockListAllPresences
+  mpaBulkPush = mockBulkPush
+  mpaStreamAdd = mockStreamAdd
+  mpaPushNative = mockPushNative
+  mpaForkIO = id  -- just don't fork.  (this *may* cause deadlocks in principle, but as long as it
+                  -- doesn't, this is good enough for testing).
+
+instance MonadNativeTargets MockGundeck where
+  mntgtLogErr _ = pure ()
+  mntgtLookupAddress = mockLookupAddress
+  mntgtMapAsync f xs = Right <$$> mapM f xs  -- (no concurrency)
+
+instance MonadPushAny MockGundeck where
+  mpyPush = mockOldSimpleWebPush
+
+instance MonadBulkPush MockGundeck where
+  mbpBulkSend = mockBulkSend
+  mbpDeleteAllPresences _ = pure ()  -- TODO: test presence deletion logic
+  mbpPosixTime = pure $ Ms 1545045904275  -- (time is constant)
+  mbpMapConcurrently = mapM  -- (no concurrency)
+  mbpMonitorBadCannons _ = pure ()  -- (no monitoring)
+
+instance Log.MonadLogger MockGundeck where
+  log _ _ = pure ()  -- (no logging)
+
+
+----------------------------------------------------------------------
+-- monad implementation
+
+-- | Expected behavior of 'Gundeck.Push.pushAll' (used in the property test).
+mockPushAll
+  :: (HasCallStack, m ~ MockGundeck)
+  => [Push] -> m ()
+mockPushAll pushes = do
+  env <- ask
+  modify $ (msWSQueue     .~ expectWS env)
+         . (msNativeQueue .~ expectNative env)
+         . (msCassQueue   .~ expectCass)
+  where
+    expectWS :: MockEnv -> NotifQueue
+    expectWS env
+      = foldl' (uncurry . deliver) mempty
+      . filter reachable
+      . reformat
+      . mconcat . fmap removeSelf
+      . mconcat . fmap insertAllClients
+      $ rcps
+      where
+        reachable :: ((UserId, ClientId), payload) -> Bool
+        reachable (ids, _) = wsReachable env ids
+
+        removeSelf :: ((UserId, Maybe ClientId, any), (Recipient, Payload)) -> [(Recipient, Payload)]
```

Comment:
It always returns exactly one value so the list and `mconcat` aren't necessary. Omission or by design?

Reply:
probably used to make sense, but doesn't any more.  fixing!